mypy/nodes.py

```diff
@@ -416,6 +416,7 @@ class FuncItem(FuncBase):
     # Is this an overload variant of function with more than one overload variant?
     is_overload = False
     is_generator = False   # Contains a yield statement?
+    is_coroutine = False   # Defined using 'async def'?
```

Comment:
I think the is_coroutine flag should correspond to the syntax used to define the function, not to the ultimate type as returned by the decorator.

See #1886 for some ramblings on that topic -- basically, the decorator should return an AwaitableGenerator, which is a new class deriving from Awaitable *and* Generator (the types I arrived at are described in https://github.com/python/mypy/issues/1886#issuecomment-233192566).